src/persist-types/src/schema.rs

```diff
@@ -0,0 +1,431 @@
+
+impl StructArrayMigration {
+    fn contains_drop(&self) -> bool {
+        use StructArrayMigration::*;
+        match self {
+            AddFieldNullable { .. } => false,
+            DropField { .. } => true,
+            AlterFieldNullable { .. } => false,
+            Recurse { migration, .. } => migration.contains_drop(),
+        }
+    }
+
+    fn name(&self) -> &String {
+        match self {
+            StructArrayMigration::AddFieldNullable { name, .. } => name,
+            StructArrayMigration::DropField { name } => name,
+            StructArrayMigration::AlterFieldNullable { name } => name,
+            StructArrayMigration::Recurse { name, .. } => name,
+        }
+    }
+
+    fn migrate(&self, len: usize, fields: &mut Fields, arrays: &mut Vec<Arc<dyn Array>>) {
+        use StructArrayMigration::*;
+        match self {
+            AddFieldNullable { name, typ } => {
+                arrays.push(new_null_array(typ, len));
+                let mut f = SchemaBuilder::from(&*fields);
+                f.push(Arc::new(Field::new(name, typ.clone(), true)));
+                *fields = f.finish().fields;
+            }
+            DropField { name } => {
+                let (idx, _) = fields
+                    .find(name)
+                    .unwrap_or_else(|| panic!("expected to find field {} in {:?}", name, fields));
+                arrays.remove(idx);
+                let mut f = SchemaBuilder::from(&*fields);
+                f.remove(idx);
+                *fields = f.finish().fields;
+            }
+            AlterFieldNullable { name } => {
+                let (idx, _) = fields
+                    .find(name)
+                    .unwrap_or_else(|| panic!("expected to find field {} in {:?}", name, fields));
+                let mut f = SchemaBuilder::from(&*fields);
+                let field = f.field_mut(idx);
+                // Defensively assert field is not nullable.
+                assert_eq!(field.is_nullable(), false);
+                *field = Arc::new(Field::new(field.name(), field.data_type().clone(), true));
+                *fields = f.finish().fields;
+            }
+            Recurse { name, migration } => {
+                let (idx, _) = fields
+                    .find(name)
+                    .unwrap_or_else(|| panic!("expected to find field {} in {:?}", name, fields));
+                arrays[idx] = migration.migrate(Arc::clone(&arrays[idx]));
+                let mut f = SchemaBuilder::from(&*fields);
+                *f.field_mut(idx) = Arc::new(Field::new(
+                    name,
+                    arrays[idx].data_type().clone(),
+                    f.field(idx).is_nullable(),
+                ));
+                *fields = f.finish().fields;
+            }
+        }
+    }
+}
+
+fn backward_compatible_typ(old: &DataType, new: &DataType) -> Option<ArrayMigration> {
+    use ArrayMigration::NoOp;
+    use DataType::*;
+    match (old, new) {
+        (Null, Null) => Some(NoOp),
```

Comment:
It seems safe and convenient to define a fallthrough `(a, b) if a == b` case (maybe restricting it to non-nested types)... would be a little more obvious to follow as a reader, and slightly more future-compatible if we end up supporting more primitive types.